geomstats/learning/riemannian_mean_shift.py

```diff
@@ -0,0 +1,161 @@
+"""
+
+import joblib
+from sklearn.base import BaseEstimator, ClusterMixin
+
+import geomstats.backend as gs
+from geomstats.learning.frechet_mean import FrechetMean
+
+
+class RiemannianMeanShift(ClusterMixin, BaseEstimator):
+    """
+    Class for Riemannian Mean Shift algorithm on manifolds.
+    Mean Shift is a procedure for locating the maxima - the modes of a
+    density function given discrete data sampled from that function. It is
+    an iterative method for finding the centers of a collection of clusters.
+
+    Following implementation assumes a flat kernel method.
+
+    Parameters
+    ----------
+    manifold : object of class RiemannianManifold
+        Geomstats Riemannian manifold on which the
+        Riemannian Mean Shift algorithm is to applied.
+    metric : object of class RiemannianMetric
+        Geomstats Riemannian metric associated to the space used.
+    bandwidth : size of neighbourhood around each center
+        All points in 'bandwidth' size around center are considered for
+        calculating new mean centers.
+    tot : float, stopping condition
+        Computation of subsequent mean centers is stopped when the distance
+        between them is less than 'tot'
+    mean : Mean method used to calculate the new center of a cluster
+    n_centers : Total initial centers
+    n_jobs : Number of parallel threads to be initiated for parallel computation
+    max_iter : int
+        Upper bound on total number of iterations for the centers to converge
+    init_centers : Initializing centers, either from the given input points or
+             random points uniformly distributed in the input manifold
+    kernel : Weighing function to assign kernel weights to each center
+    """
+
+    def __init__(
+        self,
+        manifold,
+        metric,
+        bandwidth,
+        tol=1e-2,
+        n_centers=1,
+        n_jobs=1,
+        max_iter=100,
+        init_centers="from_points",
+        kernel="flat",
+        **frechet_mean_kwargs
+    ):
+        self.manifold = manifold
+        self.metric = metric
+        self.bandwidth = bandwidth
+        self.tol = tol
+        self.mean = FrechetMean(self.metric, **frechet_mean_kwargs)
+        self.n_centers = n_centers
+        self.n_jobs = n_jobs
+        self.max_iter = max_iter
+        self.init_centers = init_centers
+        self.kernel = kernel
+        self.centers = None
+
+        print(type(max_iter))
+        print(max_iter)
+
+    def __intersets_distances(self, points_A, points_B, **joblib_kwargs):
+        """Parallel computation of distances between two sets of points"""
+        n_A, n_B = points_A.shape[0], points_B.shape[0]
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_dist(x, y):
+            """Riemannian metric between points x & y"""
+            return self.metric.dist(x, y)
+
+        pool = joblib.Parallel(n_jobs=self.n_jobs, **joblib_kwargs)
+        out = pool(
```

Comment:
You can get rid of the indices i,j here, it will improve readability.

This could be:
`pickable_dist(one_point_a, one_point_b) for one_point_a in points_a for one_point_b in points_b`